### PR TITLE
Fixed an issue with lock release in case of deleteFrom 

### PR DESCRIPTION
**Pull request by suranjan**
## Changes proposed in this pull request
In the case of deleteFrom, lock was not getting released.
Fixed the tests that were using the same session for all threads.
Added more tests and refactored code/tests

## Patch testing

precheckin
## ReleaseNotes.txt changes

(Does this change require an entry in ReleaseNotes.txt? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- store, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala [hunk 1]
```diff
@@ -262,41 +262,47 @@ case class ExecutePlan(child: SparkPlan, preAction: () => Unit = () => ())
...
+      if (shuffleIds.nonEmpty) {
+        sc.cleaner match {
+          case Some(c) => shuffleIds.foreach(c.doCleanupShuffle(_, blocking = false))
+          case None =>
+        }
+      }
+      result
+    }
+    finally {
+      logDebug(" Unlocking the table in execute of ExecutePlan ")
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala]: Could add the table name to log here. Perhaps available in session else maybe child.treeString(verbose = false)

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala [hunk 2]
```diff
@@ -278,7 +278,6 @@ abstract class BaseColumnFormatRelation(
       case None => snc.grabLock(table, schemaName, connProperties)
       case Some(a) => null // Do nothing as putInto will release lock
     }
-    if ((lock != null) && lock.isInstanceOf[RegionLock]) lock.asInstanceOf[RegionLock].lock()
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: These explicit lock acquire were redundant and grabLock itself takes the lock?

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala [hunk 3]
```diff
@@ -315,13 +314,12 @@ abstract class BaseColumnFormatRelation(
...
       case Some(_) => null // Do nothing as putInto will release lock
     }
-    if ((lock != null) && lock.isInstanceOf[RegionLock]) lock.asInstanceOf[RegionLock].lock()
     try {
       f()
     }
     finally {
-      logDebug(s"Added the ${lock} object to the context.")
       if (lock != null) {
+        logDebug(s"Added the ${lock} object to the context.")
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: Will be good to add the table being locked in these log lines

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala (hunk 2 above)
Comment (suranjan) [reply core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: Just moved the code inside the grablock.

**Pull request merged**